s2cloudless/utils.py

```diff
@@ -54,3 +71,82 @@ def get_s2_evalscript(all_bands: bool = False, reflectance: bool = False) -> str
         output_sample_type="FLOAT32" if reflectance else "UINT16",
         metadata_evalscript="" if reflectance else METADATA_EVALSCRIPT_TEMPLATE,
     ).strip("\n ")
+
+
+def get_timestamps(
+    bbox: BBox,
+    time_interval: Tuple[dt.datetime, dt.datetime],
+    *,
+    maxcc: Optional[float] = None,
+    data_collection: DataCollection = DataCollection.SENTINEL2_L1C,
+    config: Optional[SHConfig] = None,
+    time_difference: Optional[dt.timedelta] = None,
+) -> List[dt.datetime]:
+    """Get the list of timestamps for which data is available. Takes into account the bbox and time interval."""
+    time_difference = time_difference if time_difference else dt.timedelta(seconds=0)
+
+    catalog = SentinelHubCatalog(config=config)
+    cloud_cover_query = None
+    if maxcc is not None:
+        cloud_cover_query = f"eo:cloud_cover < {100 * float(maxcc)}"
+
+    search_iterator = catalog.search(
+        data_collection,
+        bbox=bbox,
+        time=time_interval,
+        filter=cloud_cover_query,
+        fields={
+            "include": [
+                "properties.datetime",
+            ],
+            "exclude": [],
+        },
+    )
+
+    return filter_times(search_iterator.get_timestamps(), time_difference)
+
+
+# pylint: disable-msg=too-many-locals
+def download_bands_and_valid_data_mask(
+    bbox: BBox,
+    timestamps: List[dt.datetime],
+    *,
+    data_collection: DataCollection = DataCollection.SENTINEL2_L1C,
+    config: Optional[SHConfig] = None,
+    size: Optional[Tuple[int, int]] = None,
+    resolution: Optional[Tuple[float, float]] = None,
+    all_bands: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Download all data required for running the cloud-masking process."""
+
+    client = SentinelHubDownloadClient(config=config)
+
+    api_requests = []
+    for timestamp in timestamps:
+        request = SentinelHubRequest(
+            evalscript=get_s2_evalscript(all_bands=all_bands, reflectance=False),
+            input_data=[
+                SentinelHubRequest.input_data(
+                    data_collection=data_collection,
+                    time_interval=(timestamp, timestamp),
+                )
+            ],
+            responses=[
+                SentinelHubRequest.output_response("bands", MimeType.TIFF),
+                SentinelHubRequest.output_response("userdata", MimeType.JSON),
+            ],
+            bbox=bbox,
+            size=size,
+            resolution=resolution,
+            config=config,
+        )
+
+        api_requests.extend(request.download_list)
+
+    responses = client.download(api_requests)
+
+    norm_factors = [response["userdata.json"]["norm_factor"] for response in responses]
+    data = np.array([response["bands.tif"] for response in responses], dtype=np.float32)
+
+    normalized_bands = np.array([np.round(array * factor, 4) for array, factor in zip(data[..., :-1], norm_factors)])
+    return normalized_bands, data[..., -1] != 0
```

Comment:
there is a lot happening in these two lines. More lines would make it more readable. 

I also don't think we benefit from the `round` since we're not saving the data anywhere. The user can do it themselves if they want.

Reply:
Round removed, calculation changed a bit, hope it is more readable.

Reply:
revert, and update ...